src/NUnitFramework/framework/Constraints/DictionaryContainsKeyConstraint.cs

```diff
@@ -66,31 +73,244 @@ public override string Description
         protected object Expected { get; }
 
         /// <summary>
-        /// Test whether the expected key is contained in the dictionary
+        /// Flag the constraint to ignore case and return self.
         /// </summary>
-        protected override bool Matches(IEnumerable actual)
+        [Obsolete(ObsoleteMessage)]
+        public new CollectionItemsEqualConstraint IgnoreCase
+        {
+            get
+            {
+                _isDeprecatedMode = true;
+                return base.IgnoreCase;
+            }
+        }
+
+        private bool Matches(object actual)
         {
-            var dictionary = ConstraintUtils.RequireActual<IDictionary>(actual, nameof(actual));
+            if (_isDeprecatedMode)
+            {
+                var dictionary = ConstraintUtils.RequireActual<IDictionary>(actual, nameof(actual));
+                foreach (object obj in dictionary.Keys)
+                    if (ItemsEqual(obj, Expected))
+                        return true;
 
-            foreach (object obj in dictionary.Keys)
-                if (ItemsEqual(obj, Expected))
-                    return true;
+                return false;
+            }
+
+            var method = GetContainsKeyMethod(actual);
+            if (method != null)
+                return (bool)method.Invoke(actual, new[] { Expected });
+
+            throw new ArgumentException("The actual value must have a ContainsKey or Contains(TKey) method.");
```

Comment:
Can we include `typeof(actual)` in this message somewhere? My reason for asking, is that I would normally hit error messages like this when I'm chaining a long list of constraints together, and I've called this constraint on a type different to the one I think I have. We've just added the type retrospectively to a couple of other constraints - so it would seem a shame to miss the opportunity on new code going into the codebase!

Reply:
@ChrisMaddock - Amended exception message for your perusal 😄 .